snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,309 @@
+                    remote_offset = subscription.get_effective_remote_consumer_group_offset(
+                        stream, self.__remote_consumer_groups
+                    )
+                    if remote_offset is None or offset >= remote_offset:
+                        self.__backend.pause([stream])
+                    elif stream not in self.__paused_streams:
+                        self.__backend.resume([stream])
+
+            if on_assign is not None:
+                on_assign(streams)
+
+        return self.__backend.subscribe(
+            topics, on_assign=assignment_callback, on_revoke=on_revoke
+        )
+
+    def unsubscribe(self) -> None:
+        # Update the subscription, so that the commit log consumer can stop
+        # consuming the commit log data for the previous (and no longer
+        # potentially assigned) topics.
+        self.__subscription.set(Subscription([]))
+
+        # TODO: How does the Kafka consumer handle paused streams? My
+        # assumption here is that the new subscription should cause those
+        # streams to be resumed again, but not sure that's accurate.
+        self.__paused_streams.clear()
+
+        return self.__backend.unsubscribe()
+
+    def poll(
+        self, timeout: Optional[float] = None
+    ) -> Optional[Message[TStream, TOffset, TValue]]:
+        # Check to see if any streams are paused that are able to be resumed
+        # based on the progression of the remote offset. (This could go the
+        # other way as well -- remote offsets may be rolled back -- but those
+        # will be caught later during normal consumption. The only benefit to
+        # pausing them here would be to avoid buffering messages that will
+        # likely be discarded. On the other hand, by the time we get to consume
+        # these messages, the offsets may have been synchronized at that point.
+        # It's hard to predict which scenario is more likely -- in the interest
+        # of buffering and low latency, it probably makes sense to avoid
+        # pausing here?)
+        with self.__subscription.get() as subscription:
+            # TODO: This could be limited to only streams that are currently
+            # paused due to the remote offsets -- we should even be able to get
+            # those streams without acquiring the lock here. (See similar note
+            # about lock-free implementation in resume.)
+            for stream, offset in self.__backend.tell().items():
+                # If a stream has been explicitly paused by the user of this
+                # class, we don't need to check the offsets: just leave it
+                # paused.
+                if stream in self.__paused_streams:
+                    continue
+
+                remote_offset = subscription.get_effective_remote_consumer_group_offset(
+                    stream, self.__remote_consumer_groups
+                )
+                if remote_offset is not None and remote_offset > offset:
```

Comment:
Is this is > and not >= because offset is the offset of the next message ?

Reply:
Yes — `offset` is the next unconsumed message offset for the local consumer, and `remote_offset` is the next unconsumed message offset for the remote consumer(s.) Same semantic as `file.tell()` providing the byte offset of the next byte to be read from a file.

It would probably be good to note this in the as-of-yet-unwritten docstring for `get_effective_remote_consumer_group_offset` even if it's not strictly necessary, I think continually reinforcing this is probably a good thing because it's easy to get tripped up because "committed offset" is easily confused with "offset of the last message processed before commit" (which are two different numbers.)